Sources/NIOFoundationCompat/ByteBuffer-foundation.swift

```diff
@@ -212,3 +212,15 @@ extension ByteBufferView: DataProtocol {
 }
 
 extension ByteBufferView: MutableDataProtocol {}
+
+// MARK: - Data
+extension Data {
+    
+    /// Creates a `Data` from a given `ByteBuffer`. The entire readable portion of the buffer will be read.
+    /// - parameter buffer: The buffer to read.
+    public init(from buffer: ByteBuffer, byteTransferStrategy: ByteBuffer.ByteTransferStrategy = .automatic) {
+        var buffer = buffer
+        self = buffer.readData(length: buffer.readableBytes, byteTransferStrategy: byteTransferStrategy)!
```

Comment:
guard with nicer error message better than force unwrap?

Reply:
I wasn't sure what to do. Having a `guard` seemed a little off as while it's safer the code *should never* execute, if we can't read `buffer.readableBytes` bytes from the buffer then something has gone horribly wrong.

Reply:
agreed it *should never* but if and when it does, do we prefer a clear error message from a guard or a cryptic one from force unwrap? are there any performance overhead in a guard?

Reply:
There'll be a minimal performance hit using a guard. What about
```
var buffer = buffer
let string = buffer.readData(length: buffer.readableBytes, byteTransferStrategy: byteTransferStrategy)!
assert(string != nil)
return string!
```

This should avoid the performance issue but still tell us something meaningful if we're in debug mode.

Reply:
Please leave the `!` and add a comment on why is is here. This will never happen and we don't need to blow up the code size for something that never happens.

Reply:
@tomerd I don't think that's a good argument. Certain things are preconditions and we don't need a message to debug. This can not and will not happen

Reply:
@weissi I believe we had similar discussion in another thread, if readData should never return nil when passing readableBytes, then having it crash instead of the call site is probably a better API design. we even mentioned the idea of adding an API which is readAllData of sorts that does not return an optional and would be used in call sites like this which would make the API design even more ergonomic. with that said, I dont think this should be holding up this PR but its a repeating theme that could/should be addressed at some point imo

Reply:
@tomerd it's impossible to do, you need the `!`. `read*` _can_ return `nil` _unless_ the parameter is `buffer.readableBytes`, in that case it cannot return `nil`. If we implemented `readAllData`, then the `!` would live inside the body of `readAllData` implementation.

 The PR we're discussing here is a PR to `swift-nio` and does what you suggests, it implements the functionality that you propose for `readAllData`, ie. never returning `nil`. Except that it uses Swift's API convention which is: Conversions from `Foo` to `Bar` are not implemented as `Foo.asBar` but rather as `Bar.init(foo: Foo)`. So consider the methods that are implemented here `readAll*`.

Reply:
So where did we land? Keep the bang?

Reply:
FWIW, I'm also with @weissi, keep the bang. It's right here.

Reply:
Alright, I'll merge as is